bibtex-actions-org-cite.el

```diff
@@ -65,6 +65,23 @@
            #'org-cite-basic--complete-style)
   :follow #'bibtex-actions-org-cite-follow)
 
+(defun bibtex-actions-org-cite-select-style (&optional keys)
+"Complete a citation style for org-cite with KEYS preview.
+
+The previews include all configured export processors, with the
+default processor highlighted with the 'bibtex-actions-org-cite-styles-default' face."
+;; TODO
+)
+
+(defun bibtex-actions-csl-render-citation (citation)
+  "Render CITATION."
+  (let ((proc (your-cached-processor)))
```

Comment:
@andras-simonyi - I added your start here. 

What is the cached-processor? I assume it's some `oc-csl` variable?

Reply:
I meant the `citeproc-el` citation processor used for rendering the previews. It makes sense to use a cached one to avoid  reloading the used CSL style unnecessarily. It should be indeed the value of a variable (as in `org-cite-csl-activate`), and the safest option would be to use a buffer-local one which is dedicated to this task, i.e., rendering previews. Thinking about it a bit more, actually it could be shared with `org-cite-csl-activate`, but we probably don't want to engage in this kind of premature optimization at this stage :-).

Reply:
> but we probably don't want to indulge in this kind of premature optimization at this stage

So bottomline, it's not shared with `oc-csl`?

Therefore I would create one here?

Reply:
I think so. An even more lighweight option would be not using a full-fledged processor but just the style object and `citeproc-render-item`  (see the "Rendering isolated references" section in the README. Tomorrow I can post some code along these lines.

Reply:
> I'd copy the solution in `org-cite-csl-activate`, and later think about whether more sharing/reuse can be achieved.

I haven't tested it (need to hook it to annotations, etc.), but should be done. Thanks!

Reply:
So @andras-simonyi, I have basic grouping annotation working, but in looking at the CSL angle I'm a little unsure.

I need to use the org-cite-make-insert function, where the insert-style argument to hook my function up. But that function is defined as taking 0 arguments. So I don't see how to get the citation to preview.

Maybe I can just select one at random?

Reply:
I'm not sure about the details but my hunch is that the ideal solution would be work with the citation object at point, which you can get, I think, using `org-element-context`. (You might find it useful to look at  `org-cite-csl-activate--sensor-fun`, `element` is the citation object to be rendered in the code.) If you manage to get hold of the citation object, then you can probably generate from it all the required stylistic variants and render them.

Reply:
> ... my hunch is that the ideal solution would be working with the citation object at point, which you can get, I think, using `org-element-context`. (You might find it useful to look at `org-cite-csl-activate--sensor-fun`, `element` is the citation object to be rendered in the code.) If you manage to get hold of the citation object then you can probably generate from it all the required stylistic variants and render them.

Right, but I can't see how to do that ATM; from _within_ the `SELECT-STYLE` function.

So let's review, in part so I can think this through while writing it.

The annotation function, where I am putting these previews, takes a candidate as input, and returns some (propertized) string.

The candidate is a style+variant string.

The `bibtex-actions-org-cite-select-style` function handles the UI, but cannot (at least ATM) take any arguments; so unlike some org-cite functions, it cannot take the citation as argument.

Here's the code that constructs the insert processor. 

```elisp
(org-cite-register-processor 'bibtex-actions-org-cite
  :insert (org-cite-make-insert-processor
           #'bibtex-actions-org-cite-insert
           #'bibtex-actions-org-cite-select-style) ; this what I'm changing here, but this can't take any args
  :follow #'bibtex-actions-org-cite-follow)
```
To be clear, I don't have to use `org-cite-make-insert-processor`; it's just easier, and I'm trying to keep this simple as possible :-)

`Org-cite-insert`, which this is using, is context aware, but I'm not seeing how the `SELECT-STYLE` function can know about that.

So the issue is how to get the citation within that `bibtex-actions-org-cite-select-style` function.

If I can do that, not only will it be a better UX, but it will drastically reduce the amount of code here, because I can use the export processors for all the previews, rather than the alists I'm currently using, and treat the csl and tex export processors the same.

I was thinking about maybe having some cache of previews that is constantly updated, but not sure.

Or maybe I'm missing something obvious, which is why I asked.

cc  @jkitchin, in case he has any ideas, since i know in his experiments, he is not using `org-cite-make-insert-processor`, but also had to write a lot more code for this.

https://github.com/jkitchin/scimax/blob/3ce9942ddadb09176d0fae8db161cf3f5324766a/org-ref3/oc-bibtex.el#L336

Reply:
PS - I posted a simplified version of this question to the list. I think that's the best solution, if I'm understanding correctly.

Reply:
Here are a few reasons my current code doesn't use `org-cite-make-insert-processor`:

0. It feels complicated to me to use a function that takes two functions and then returns a function; I found it hard to debug, so I wrote out a full function I could debug.
 
1. Mostly it is because (I think) it is not compatible with the way ivy does sequential actions. This needs to call a function with the selected candidate as the arg. To solve that, I separated the insert into two functions, one that uses ivy to select, with an action that calls a second function that looks more like the  one created by `org-cite-make-insert-processor`. This is probably specific to ivy, and would not apply the same to helm, or probably other completion tools.  

2. Second, as @bdarcus discusses above, there are places where I need a new context, because the citation has changed since `org-insert-cite` was initially called, e.g. due to an insertion. I get the context inside my insert function on every call.

Reply:
Thanks @jkitchin. Yes, I can see your 0 point.

I may end up going that path as well, though will see if Nicolas pops his head up and has a response.

On this:

> Mostly it is because (I think) it is not compatible with the way ivy does sequential actions. 

Was this because you wanted to enable moving and sorting of multi-cites?

I was actually wondering how to do that in just `completing-read`/`completing-read-multiple`, but hadn't thought about it deeply yet. It may be possible, but am not sure.

Reply:
With ivy, the following options exist for sequential inserts (without closing the selection buffer), is you type C-M-m. That will call the default action with the current selection, but it does not close the selection buffer.  If you want an alternate action, you type C-M-o, select the action, press Ret, and then you go back to the selection buffer. There is an alternative way to mark candidates, and then the action is applied to each candidate in the marked list. Neither of these  approaches works well with the way `org-cite-insert` was working.

The moving/sorting is all done by a keymap that goes on in the activate function, or by functions that work at point in the follow hydra.

Reply:
> With ivy, the following options exist for sequential inserts (without closing the selection buffer), is you type C-M-m. That will call the default action with the current selection, but it does not close the selection buffer.

Because ivy doesn't have support for multi-selection a la helm?

In bibtex-actions, I use `completing-read-multiple` for that.

> The moving/sorting is all done by a keymap that goes on in the activate function ...

I'm curious now; will take a look.

Reply:
ivy does do something kind of like helm where you can mark a set of candidates, and then apply a function to each candidate. It isn't quite like helm though. In the beginning of ivy though, it was not like helm, and the ability to mark things came later.

Reply:
> 1. I found it hard to debug, so I wrote out a full function I could debug.

This bit me hard yesterday. I think at minimum that function should have a more helpful error message, to make clear where the problem is.

I ran into trouble when I tried to implement support for multiple previews. For now, I've decided to remove that, and just allow users to choose one option.

Also, the previews are not generated; I just have an alist with them, including the start for CSL, which I am manually creating.

I will probably merge this as is, but I think ideally we want:

1. possibility for multiple preview targets (user-configurable, with maybe default being automatically derived from the user org-cite export processor config)
2.  the actual style candidates should be extracted from `org-cite-supported-styles` (it's not currently)
2. related, rather than using an alist, just use the export processors to generate the previews, ideally with the keys at point. That's less code and more general (works the same for tex and for csl).

I was also wondering again if the insert processor(s) should be a separate, possibly collaborative, project.

I mean, I want to use completing-read with this, and to be able to plugin bibtex-actions for the "select-keys" functionality, but otherwise, it's conceptually independent of this project, which is really just a completing-read front-end to bibtex-completion.

Reply:
> > 1. I found it hard to debug, so I wrote out a full function I could debug.
> 
> This bit me hard yesterday. I think at minimum that function should have a more helpful error message, to make clear where the problem is.

I fixed it by placing `org-cite-make-insert-processor` at end of the file.